core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -555,24 +479,43 @@ private object RTS {
                       io.value.ensuring(exitSupervision(io.supervisor))
 
                   case IO.Tags.Terminate =>
-                    val io = curIo.asInstanceOf[IO.Terminate]
+                    val io = curIo.asInstanceOf[IO.Terminate[E]]
 
-                    val finalizer = interruptStack
+                    val finalizer = interruptStack(io.cause.checkedFirst.isDefined)
 
-                    if (finalizer eq null) {
-                      // No finalizers, simply produce error:
-                      val causes    = status.get.terminationCauses.getOrElse(Nil)
-                      val defects   = status.get.defects
-                      val allCauses = io.causes ++ causes ++ defects
+                    if (stack.isEmpty) {
+                      // Error not caught, stack is empty:
+                      if (finalizer eq null) {
+                        // No finalizer, so immediately produce the error.
+                        curIo = null
+
+                        val cause = (io.cause, status.get.terminationCause) match {
+                          case (Interruption(None), Some(cause)) if cause.interruptions.isDefined => cause
+                          case _                                                                  => status.get.terminationCause.fold(io.cause)(io.cause ++ _)
+                        }
+
+                        result = ExitResult.Failed(cause)
+                      } else {
+                        // We have finalizers to run. We'll resume executing with the
+                        // uncaught failure after we have executed all the finalizers:
+                        val finalization = finalizer.flatMap(accumFailures)
+                        val completer    = io
 
-                      curIo = null
-                      result = ExitResult.Terminated(allCauses)
+                        curIo = doNotInterrupt(finalization) *> completer
+                      }
                     } else {
-                      // Must run finalizers first before failing:
-                      val finalization = finalizer.flatMap(accumFailures)
-                      val completer    = io
+                      // Error caught:
+                      val handled = nextInstr[E](io.cause, stack)
+
+                      if (finalizer eq null) {
+                        curIo = handled
+                      } else {
+                        // Must run finalizer first:
+                        val finalization = finalizer.flatMap(accumFailures)
```

Comment:
This part requires more intricate surgery. Right now:

1. `handled` is created to handle the `io.cause`. But that's wrong, because we need to insert the finalizer failures into `io.cause`, before we can create `handled`.
2. The finalizer failures are being accumulated into the fiber status but not really used, which means they'll sit around until the fiber finishes (I assume).

I think the new code for this section should be something like:

```scala
if (finalizer eq null) {
  curIo = nextInstr[E](io.cause, stack)
} else {
  curIo = doNotInterrupt(finalizer).map(cause => io.cause ++ cause)
}
```

Then after the new cause is computed (as the concat of the `io.cause` and the finalizer failure causes), then the next operation to be popped off the stack will be `nextInstr[E](io.cause ++ [finalizer cause])` so the handler will get the full cause.